### ISSUE TITLE
Invalid html produced from our default HTML Pipeline

### ISSUE THREAD

**Issue by timdiggins** [thredded/thredded#941]
Looking at #940 ,  I realised that while the markdown content:

```markdown
**http://www.example.com**
```

will generate something anodyne like:

```html
<!DOCTYPE html>
<title>whatever</title>
<body>
<strong><a href="http://www.example.com"></strong>
</body>
```

which is fully valid.

But with onebox enabled , then the following markdown:

```markdown
**https://github.com/thredded/thredded/commit/7276f52e17e2bc80b1fe49e8894bf35c4e2e724a**
```

will result in a complex structure like:

```html
<!DOCTYPE html>
<title>whatever</title>
<body>
  </strong><strong>
    <header class="source">...clipped...</header>
    <article class="onebox-body">...clipped...</article>
    <div class="onebox-metadata"></div>
    <div style="clear: both"></div>
  </strong>
</body>
```

This is invalid html5  because flow elements can't be contained inside a phrasing element like `<strong>`(https://html.spec.whatwg.org/multipage/text-level-semantics.html#the-strong-element). A oneboxed result will I think always generate flow elements not purely phrasing elements. 

Browsers will handle this, but shouldn't we be somehow fixing this? -- maybe not oneboxing if it's a phrasing element? or dropping the phrasing? Sounds complicated either way.

Thoughts @glebm ?

**Comment by glebm** [comment]
We have `extract_block_from_paragraph!` which works around a similar issue for paragraphs.

Perhaps we can add something similar for bolding/italics/underscore?

**Comment by timdiggins** [comment]
Do you think maybe we shouldn't attempt to perform onebox if it's inside a bold / italic/ underscore. 

That might mean adapting the following line:

https://github.com/thredded/thredded/blob/faeda40782221d07b5e4126d6577f3849ccef193/lib/thredded/html_pipeline/onebox_filter.rb#L51

and replace `on_its_own_line?` with a new method `allows_flow_content?` (which can just look at what tag the parent element is). 

Sound plausible?

**Comment by glebm** [comment]
Yes, that makes sense. It should still be allowed within paragraphs though.
